static/js/ace-editor.ts

```diff
@@ -1,230 +1,185 @@
 import { Markers } from "./markers";
+import { EventEmitter } from "./event-emitter";
 // const MOVE_CURSOR_TO_BEGIN = -1;
 const MOVE_CURSOR_TO_END = 1;
 export class HedyAceEditorCreator implements HedyEditorCreator {
   /**
    * This function should initialize the editor and set up all the required
    * event handlers
    * @param $editor reference to the div that contains the main editor
+   * @param editorType the type of the editor
+   * @param dir the direction of the text
+   * @return {HedyEditor} The initialized Hedy editor instance
    */
-  initializeMainEditor($editor: JQuery): HedyAceEditor {
-    let editor: HedyAceEditor = this.turnIntoEditor($editor.get(0)!, $editor.data('readonly'), true);
-    editor.configureMainEditor();
-    error.setEditor(editor);
-    window.Range = ace.require('ace/range').Range // get reference to ace/range
-
-    // *** KEYBOARD SHORTCUTS ***
-    let altPressed: boolean | undefined;
-    // alt is 18, enter is 13
-    window.addEventListener('keydown', function (ev) {
-      const keyCode = ev.keyCode;
-      if (keyCode === 18) {
-        altPressed = true;
-        return;
-      }
-      if (keyCode === 13 && altPressed) {
-        if (!theLevel || !theLanguage) {
-          throw new Error('Oh no');
-        }
-        runit(theLevel, theLanguage, "", function () {
-          $('#output').focus();
-        });
-      }
-      // We don't use jquery because it doesn't return true for this equality check.
-      if (keyCode === 37 && document.activeElement === document.getElementById('output')) {
-        editor.focus();
-        editor.moveCursorToEndOfFile();
-      }
-    });
-
-    window.addEventListener('keyup', function (ev) {
-      triggerAutomaticSave();
-      const keyCode = ev.keyCode;
-      if (keyCode === 18) {
-        altPressed = false;
-        return;
-      }
-    });
-
+  initializeWritableEditor($editor: JQuery, editorType: EditorType, dir: string = "ltr"): HedyAceEditor {
+    let editor: HedyAceEditor = new HedyAceEditor($editor.get(0)!, $editor.data('readonly'), editorType, dir);
     return editor;
   }
+  /**
+   * Initializes a read only editor
+   *
+   * @param {HTMLElement} preview - The element to preview the editor.
+   * @return {HedyEditor} The initialized Hedy editor instance.
+   */
+  initializeReadOnlyEditor(preview: HTMLElement, dir: string = "ltr"): HedyEditor {
+    let editorType: EditorType;
+  
+    if ($(preview).hasClass('common-mistakes')) {
+      editorType = EditorType.COMMON_MISTAKES;
+    } else if ($(preview).hasClass('cheatsheet')) {
+      editorType = EditorType.CHEATSHEET;
+    } else if ($(preview).hasClass('parsons')) {
+      editorType = EditorType.PARSONS;
+    } else {
+      editorType = EditorType.EXAMPLE;
+    }
+
+    return new HedyAceEditor(preview, true, editorType, dir);
+  }
+}
+  
+export class HedyAceEditor implements HedyEditor {
+  private _editor: AceAjax.Editor;
+  private _markers?: Markers
+  isReadOnly: boolean;
```

Comment:
You probably want to make this 

```suggestion
  private isReadOnly: boolean;
```

Because right now in other parts of the code you're accidentally directly accessing this public field, instead of going through the getter/setter as I think you probably intended.